client/README.md

```diff
@@ -1,208 +1,389 @@
+  - [Handling Macro/Templating Languages](#handling-macrotemplating-languages)
+  - [Custom Pre and Postprocessing Logic](#custom-pre-and-postprocessing-logic)
+- [Renaming SQL Identifiers](#renaming-sql-identifiers)
+- [Deploying to Cloud Run](#deploying-to-cloud-run)
+
+## Quickstart
+
+```shell
+# Install the CLI as described in the Installation section below.
+
+# Copy the example project directory to a project directory of your own.
+cp -R examples/teradata/sql <YOUR_PROJECT_DIRECTORY>
+
+# Change directory to your project directory.
+cd <YOUR_PROJECT_DIRECTORY>
+
+# Remove the example input files from the input directory.
+rm -rf input/*
+
+# Copy the files you would like to translate into the input directory.
+cp <YOUR_INPUT_FILES> input/
+
+# Edit the config/config.yaml file appropriately as described in the Basic Usage
+# section below.
+nano config/config.yaml
+
+# Set the appropriate environment variables as described in the Basic Usage
+# section below.
+export BQMS_VERBOSE="False"
+export BQMS_MULTITHREADED="True"
+export BQMS_PROJECT="<YOUR_GCP_PROJECT>"
+export BQMS_GCS_BUCKET="<YOUR_GCS_BUCKET>"
+
+# Run the translation.
+./run.sh
 ```
-pip3 install dwh-migration-tools/client
-```
-
-Install the gcloud CLI following the [instructions](http://cloud.google.com/sdk/docs/install).
-
-### [Optional] gcloud login and authentication
 
-The program will first validate the login and credential status of
-gcloud. If the validation steps failed, the program will run the following two
-commands automatically and bring up a page on browser for your agreement of
-using your Google account.
+## Installation
 
-However, it's recommended that you run these two commands on the terminal first as a one-time setup requirement:
+Prerequisites: Python >= 3.7.2.
 
-Log in to gcloud:
+Preferred OS: Linux or MacOS.
 
-```
-gcloud auth login
+```shell
+pip install .
 ```
 
-Generate an application-default credential file so that you can use gcloud API
-programmatically:
-
-```
-gcloud auth application-default login
+## Basic Usage
+
+There are two ways to run the CLI: directly and via a wrapping  `run.sh` script.
+
+Direct execution requires setting
+[environment variables](#environment-variables) and modifying a
+[config.yaml](#configyaml) file.
+
+`run.sh` execution is described in detail in 
+[Recommended Usage](#recommended-usage).
+
+### Environment Variables
+
+The CLI accepts the following environment variables:
+
+  - `BQMS_PROJECT` **(required)**: The project that the translation job will be
+    run in. All GCS paths must reside in this project.
+  - `BQMS_INPUT_PATH` **(required)**: The path that raw input to be preprocessed
+    will be read from . This can either be a local path or a path that 
+    begins with the `gs://` scheme.
+  - `BQMS_PREPROCESSED_PATH` **(required)**: The path that the batch SQL
+    translator will read preprocessed input from. It must be a path that begins
+    with the `gs://` scheme and thus always reside on GCS.
```

Comment:
I have an impression that the cloud-run option won't change the logic of local run but it seems it is no longer the case.
Personally I would prefer the option of local run without the need to upload/download the temporary files to/from GCS.
For existing local run users, they will probably see a significant latency increase now if they have lots of input files.

Reply:
> I have an impression that the cloud-run option won't change the logic of local run but it seems it is no longer the case.

No you are correct, the logic is the same.

`BQMS_PREPROCESSED_PATH` and `BQMS_TRANSLATED_PATH` have to be on GCS because BQMS only understands GCS for IO. This is the case for both local and Cloud Run execution. There is no added latency or additional uploads/downloads in Python above and beyond what the current client does for the local execution case. In Python, there's one upload and one download per file.

The `run.sh` has one final step that downloads what used to be considered "temp files" i.e the input/output to/from BQMS. If we want to view the result of preprocessing (i.e. the input to BQMS) without GCS, I think we should favor @Hackerpilot 's approach in #112 of a standalone command.

Reply:
Just want to add an update that I plan to add a performance test in the next couple days so that we have hard data to work off of and we can ensure this matches the performance of the existing client.

Reply:
I added a performance test as outlined [here](https://github.com/google/dwh-migration-tools/blob/731011bdc2fc8adcd0fd08a28147e464246ef023/client/README.md?plain=1#L393-L538).

The performance test for the existing client is [here](https://github.com/google/dwh-migration-tools/compare/main...ajwelch4:dwh-migration-tools:perf-test).

My machine:

```shell
> free -hm
               total        used        free      shared  buff/cache   available
Mem:            94Gi        10Gi        63Gi       146Mi        20Gi        82Gi
Swap:          1.9Gi        32Mi       1.8Gi

> lscpu

Architecture:            x86_64
  CPU op-mode(s):        32-bit, 64-bit
  Address sizes:         46 bits physical, 48 bits virtual
  Byte Order:            Little Endian
CPU(s):                  24
  On-line CPU(s) list:   0-23
Vendor ID:               GenuineIntel
  Model name:            Intel(R) Xeon(R) CPU @ 2.20GHz
    CPU family:          6
    Model:               79
    Thread(s) per core:  2
    Core(s) per socket:  12
    Socket(s):           1
    Stepping:            0
    BogoMIPS:            4400.41
    Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc
                          cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single pti ssbd ibrs
                          ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat md_clear arch_capabilities
Virtualization features: 
  Hypervisor vendor:     KVM
  Virtualization type:   full
Caches (sum of all):     
  L1d:                   384 KiB (12 instances)
  L1i:                   384 KiB (12 instances)
  L2:                    3 MiB (12 instances)
  L3:                    55 MiB (1 instance)
NUMA:                    
  NUMA node(s):          1
  NUMA node0 CPU(s):     0-23
```

Results:

### 10k files * 10 KB/file = 100 MB of data

|                                                 | Client         | Service       | Total                |
|-----------------------------|------------|------------|---------------|
| Existing client                        | 0h 17m 11s | 0h 08m 25s | 0h 25m 36s |
| This client, single-threaded | 0h 17m 36s | 0h 08m 30s | 0h 26m 06s |
| This client, multi-threaded   | 0h 02m 17s | 0h 11m 56s | 0h 14m 13s |

### 100k files * 10 KB/file = 1 GB of data (service limit)

|                                                 | Client          | Service          | Total             |
|-----------------------------|------------|--------------|-------------|
| Existing client                        | 2h 31m 11s | 1h 21m 24s | 3h 52m 35s |
| This client, single-threaded | 2h 35m 39s | 1h 40m 00s | 4h 15m 39s |
| This client, multi-threaded   | 0h 21m 02s | 1h 24m 44s | 1h 45m 46s |